File: src/deals_api.py
"""
Deal Awareness API - "Where & When" Flow

Endpoints:
- GET /api/deals/evaluate - Evaluate deal quality for a route/month
- GET /api/deals/batch - Batch evaluate multiple routes

Uses Supabase RPC to call the evaluate_deal() PostgreSQL function.
"""

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field
from typing import Optional, List
import os
import httpx
from datetime import datetime

router = APIRouter(prefix="/api/deals", tags=["deals"])

# Environment
SUPABASE_URL = os.environ.get("SUPABASE_URL", "").rstrip("/")
SUPABASE_SERVICE_ROLE = os.environ.get("SUPABASE_SERVICE_ROLE", "")

if not SUPABASE_URL or not SUPABASE_SERVICE_ROLE:
    print("⚠️  WARNING: SUPABASE_URL or SUPABASE_SERVICE_ROLE not set. /api/deals/* will fail.")


# Response models
class Baseline(BaseModel):
    p25: float = Field(..., description="25th percentile price (30-day rolling)")
    p50: float = Field(..., description="50th percentile (median) price")
    p75: float = Field(..., description="75th percentile price")
    samples: int = Field(..., description="Number of price samples in baseline")
    last_updated: Optional[datetime] = Field(None, description="When baseline was last computed")


class SweetSpot(BaseModel):
    min_days: int = Field(..., description="Minimum lead time in sweet-spot window")
    max_days: int = Field(..., description="Maximum lead time in sweet-spot window")


class DealEvaluation(BaseModel):
    has_data: bool = Field(..., description="Whether sufficient data exists for evaluation")
    message: Optional[str] = Field(None, description="Error or warning message")
    origin: str
    dest: str
    month: int = Field(..., ge=1, le=12, description="Departure month (1-12)")
    cabin: str = Field(default="economy")
    depart_month: Optional[str] = Field(None, description="Actual departure month date (YYYY-MM-DD)")
    current_price: Optional[float] = Field(None, description="Current cheapest price for this route/month")
    baseline: Optional[Baseline] = Field(None, description="30-day rolling baseline")
    delta_pct: Optional[float] = Field(None, description="% difference from median baseline")
    deal_score: Optional[int] = Field(None, ge=0, le=100, description="Deal quality score (0-100)")
    sweet_spot: Optional[SweetSpot] = Field(None, description="Optimal booking window")
    recommendation: Optional[str] = Field(None, description="BUY, TRACK, or WAIT")
    confidence: Optional[int] = Field(None, ge=0, le=100, description="Recommendation confidence %")
    rationale: Optional[str] = Field(None, description="Human-readable explanation")
    last_seen: Optional[datetime] = Field(None, description="When current price was last observed")


class BatchEvaluateRequest(BaseModel):
    routes: List[dict] = Field(..., description="List of {origin, dest, month, cabin?} objects")


# Endpoints
@router.get("/evaluate", response_model=DealEvaluation)
async def evaluate_deal(
    origin: str = Query(..., min_length=3, max_length=3, description="Origin airport code (e.g., JFK)"),
    dest: str = Query(..., min_length=3, max_length=3, description="Destination airport code (e.g., MIA)"),
    month: int = Query(..., ge=1, le=12, description="Departure month (1-12)"),
    cabin: str = Query("economy", description="Cabin class (economy, premium, business, first)")
):
    """
    Evaluate deal quality for a specific route and month.

    Returns:
    - BUY: Excellent deal, book now
    - TRACK: Fair deal, monitor for improvements
    - WAIT: Poor deal, wait for prices to drop

    Includes:
    - Current price vs 30-day baseline (P25/P50/P75)
    - Deal score (0-100)
    - Sweet-spot booking window (if available)
    - Confidence level and rationale
    """
    if not SUPABASE_URL or not SUPABASE_SERVICE_ROLE:
        raise HTTPException(
            status_code=500,
            detail="Supabase not configured. Set SUPABASE_URL and SUPABASE_SERVICE_ROLE."
        )

    # Call Supabase RPC endpoint
    url = f"{SUPABASE_URL}/rest/v1/rpc/evaluate_deal"
    headers = {
        "apikey": SUPABASE_SERVICE_ROLE,
        "Authorization": f"Bearer {SUPABASE_SERVICE_ROLE}",
        "Content-Type": "application/json",
        "Prefer": "return=representation"
    }
    payload = {
        "p_origin": origin.upper(),
        "p_dest": dest.upper(),
        "p_month": month,
        "p_cabin": cabin.lower()
    }

    try:
        async with httpx.AsyncClient(timeout=20.0) as client:
            response = await client.post(url, headers=headers, json=payload)

        if response.status_code >= 400:
            raise HTTPException(
                status_code=502,
                detail={
                    "error": "Supabase RPC failed",
                    "status": response.status_code,
                    "message": response.text
                }
            )

        data = response.json()

        # Supabase RPC may return array or object depending on settings
        if isinstance(data, list) and len(data) > 0:
            data = data[0]

        return DealEvaluation(**data)

    except httpx.TimeoutException:
        raise HTTPException(status_code=504, detail="Request to Supabase timed out")
    except httpx.RequestError as e:
        raise HTTPException(status_code=502, detail=f"Failed to connect to Supabase: {str(e)}")
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Deal evaluation failed: {str(e)}")


@router.post("/batch", response_model=List[DealEvaluation])
async def batch_evaluate(request: BatchEvaluateRequest):
    """
    Batch evaluate multiple routes at once.

    Request body:
    {
      "routes": [
        {"origin": "JFK", "dest": "MIA", "month": 3},
        {"origin": "JFK", "dest": "LAX", "month": 6, "cabin": "business"}
      ]
    }

    Returns array of DealEvaluation objects.
    """
    results = []

    for route in request.routes:
        try:
            origin = route.get("origin", "")
            dest = route.get("dest", "")
            month = route.get("month", 1)
            cabin = route.get("cabin", "economy")

            eval_result = await evaluate_deal(origin, dest, month, cabin)
            results.append(eval_result)

        except HTTPException as e:
            # Include error as failed evaluation
            results.append(
                DealEvaluation(
                    has_data=False,
                    message=str(e.detail),
                    origin=route.get("origin", ""),
                    dest=route.get("dest", ""),
                    month=route.get("month", 1),
                    cabin=route.get("cabin", "economy")
                )
            )

    return results


@router.get("/health")
async def deals_health():
    """Check if deals API is operational."""
    if not SUPABASE_URL or not SUPABASE_SERVICE_ROLE:
        return {
            "status": "unhealthy",
            "message": "Supabase credentials not configured"
        }

    return {
        "status": "healthy",
        "supabase_url": SUPABASE_URL.replace(SUPABASE_URL.split("//")[1].split(".")[0], "***"),
        "features": ["evaluate", "batch"]
    }

File: v4/frontend/src/lib/dealsApi.ts
/**
 * Deal Awareness API Client
 *
 * Provides functions to evaluate deal quality and get booking recommendations.
 * Connects to /api/deals/* endpoints.
 */

// Get API base from config or environment
export const API_BASE =
  (window as any).__APP_CONFIG__?.VITE_API_BASE ||
  import.meta.env.VITE_API_BASE ||
  'http://localhost:8000';

/**
 * Baseline price distribution (30-day rolling)
 */
export interface Baseline {
  p25: number;        // 25th percentile
  p50: number;        // 50th percentile (median)
  p75: number;        // 75th percentile
  samples: number;    // Number of price observations
  last_updated?: string;
}

/**
 * Optimal booking window (sweet-spot)
 */
export interface SweetSpot {
  min_days: number;   // Minimum lead time
  max_days: number;   // Maximum lead time
}

/**
 * Deal evaluation response
 */
export interface DealEvaluation {
  has_data: boolean;
  message?: string;
  origin: string;
  dest: string;
  month: number;               // 1-12
  cabin: string;
  depart_month?: string;       // YYYY-MM-DD
  current_price?: number;
  baseline?: Baseline;
  delta_pct?: number;          // % vs median baseline
  deal_score?: number;         // 0-100
  sweet_spot?: SweetSpot | null;
  recommendation?: 'BUY' | 'TRACK' | 'WAIT';
  confidence?: number;         // 0-100
  rationale?: string;
  last_seen?: string;
}

/**
 * Evaluate deal quality for a route and month
 *
 * @param origin - Origin airport code (e.g., "JFK")
 * @param dest - Destination airport code (e.g., "MIA")
 * @param month - Departure month (1-12)
 * @param cabin - Cabin class (default: "economy")
 * @returns Deal evaluation with BUY/TRACK/WAIT recommendation
 */
export async function evaluateDeal(
  origin: string,
  dest: string,
  month: number,
  cabin: string = 'economy'
): Promise<DealEvaluation> {
  const url = new URL(`${API_BASE}/api/deals/evaluate`);
  url.searchParams.set('origin', origin.toUpperCase());
  url.searchParams.set('dest', dest.toUpperCase());
  url.searchParams.set('month', String(month));
  url.searchParams.set('cabin', cabin.toLowerCase());

  const response = await fetch(url.toString());

  if (!response.ok) {
    const error = await response.json().catch(() => ({ message: response.statusText }));
    throw new Error(error.message || `Deal evaluation failed: ${response.status}`);
  }

  return response.json();
}

/**
 * Batch evaluate multiple routes
 *
 * @param routes - Array of {origin, dest, month, cabin?} objects
 * @returns Array of deal evaluations
 */
export async function batchEvaluateDeal(
  routes: Array<{
    origin: string;
    dest: string;
    month: number;
    cabin?: string;
  }>
): Promise<DealEvaluation[]> {
  const response = await fetch(`${API_BASE}/api/deals/batch`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ routes })
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({ message: response.statusText }));
    throw new Error(error.message || `Batch evaluation failed: ${response.status}`);
  }

  return response.json();
}

/**
 * Get recommendation color based on type
 */
export function getRecommendationColor(recommendation?: string): string {
  switch (recommendation) {
    case 'BUY':
      return 'text-glow-lime';
    case 'WAIT':
      return 'text-glow-magenta';
    case 'TRACK':
      return 'text-note-cyan';
    default:
      return 'text-text-dim';
  }
}

/**
 * Get recommendation badge variant
 */
export function getRecommendationBadge(recommendation?: string): string {
  switch (recommendation) {
    case 'BUY':
      return 'bg-glow-lime/20 text-glow-lime border-glow-lime';
    case 'WAIT':
      return 'bg-glow-magenta/20 text-glow-magenta border-glow-magenta';
    case 'TRACK':
      return 'bg-note-cyan/20 text-note-cyan border-note-cyan';
    default:
      return 'bg-ink-700 text-text-dim border-ink-500';
  }
}

/**
 * Format delta percentage with sign and color
 */
export function formatDelta(delta_pct?: number): {
  text: string;
  color: string;
} {
  if (delta_pct === undefined || delta_pct === null) {
    return { text: 'N/A', color: 'text-text-dim' };
  }

  const sign = delta_pct > 0 ? '+' : '';
  const color = delta_pct <= 0 ? 'text-glow-lime' : 'text-glow-magenta';

  return {
    text: `${sign}${delta_pct.toFixed(1)}%`,
    color
  };
}

/**
 * Get deal score color based on value
 */
export function getDealScoreColor(score?: number): string {
  if (!score) return 'text-text-dim';
  if (score >= 80) return 'text-glow-lime';
  if (score >= 60) return 'text-note-cyan';
  if (score >= 40) return 'text-glow-yellow';
  return 'text-glow-magenta';
}
